### ISSUE TITLE
Unable to query from a xsd:decimal literal (on the top of a DB2)

### ISSUE THREAD

**Issue by marcelomachado** [ontop/ontop#749]
### Description
I am mapping a DB2 Table to RDF. This table has a column (`column_name`) of type double. Here's an R2RML snippet of that mapping:
```
:pom_042 rdf:type rr:PredicateObjectMap ;
	rr:predicateMap :pm_042 ;
	rr:objectMap :om_042 .

:pm_042 rdf:type rr:PredicateMap ;
	rr:constant ex:some_predicate .

:om_042 rdf:type rr:ObjectMap ;
	rr:column "column_name" ;
	rr:termType rr:Literal ;
	rr:datatype xsd:decimal .
```

```
SELECT ?o WHERE
{
    ?s ex:some_predicate ?o
}
LIMIT 1
```
The result is like: `"259.28"^^xsd:decimal` for `o`. But, when I replace `?o` for  `"259.28"^^xsd:decimal` I get no results. I mean, the following query returns an empty result for `?s`:
```
SELECT ?s WHERE
{
    ?s ex:some_predicate "259.28"^^xsd:decimal`
}
LIMIT 1
```
Even stranger, when I do not set a predicate in that SPARQL the result returned for `o` is `"2.5928E2"^^xsd:decimal` (with this `E2` at the end):

```
SELECT * WHERE
{
    ex:some_specific_subject ?p ?o
}
LIMIT 1
```

What's going on and how can I make a query using a decimal?

### Steps to Reproduce

**Expected behavior:**
The following query should return some result for `s`:
```
SELECT ?s WHERE
{
    ?s ex:some_predicate "259.28"^^xsd:decimal`
}
LIMIT 1
```

**Actual behavior:**
The following query returns nothing:
```
SELECT ?s WHERE
{
    ?s ex:some_predicate "259.28"^^xsd:decimal`
}
LIMIT 1
```

**Reproduces how often:** 
Yes

### Versions

Which version of Ontop are you using and in which setting (Protege, SPARQL endpoint, Cli, Java API, ...)

I am using [the latest Docker image  of the Ontop endpoint](https://hub.docker.com/layers/ontop/ontop-endpoint/latest/images/sha256-e4baf868418947dbbe251e796aa1658890dee7973a1d8139fe5132cb6f457cf9?context=explore).

DB2 version: DB2 v11.1.4.7

DB2 JDBC Driver:
https://mvnrepository.com/artifact/com.ibm.db2.jcc/db2jcc/db2jcc4

**Comment by bcogrel** [comment]
Hi @marcelomachado, thanks for reporting the issue.

We need to be more consistent with the normalization and denormalization of decimal lexical values with DB2. The engineering notation should actually not be used for `xsd:decimal`, we need to avoid it.

Instead of the comparing the exact lexical values, have you considered using a FILTER clause?

```sparql
SELECT ?s WHERE {
    ?s ex:some_predicate ?value)
   FILTER (?value = "259.28"^^xsd:decimal)
}
LIMIT 1
```
In this case, it involves a numeric equality, which is less strict than lexical value equality.

Best,
Benjamin

**Comment by marcelomachado** [comment]
Hello @bcogrel, thanks for answering.

Yes, I considered that, and for now it's the solution I'm using. However, this comparison brings results beyond the expected considering the precision point. For example, some results returned by this query: 259.33, 259.298, 259.102.

That is the SQL query created:
```sql
SELECT V1."column_name" AS "column_name 1m23"
FROM "MY_SCHEMA"."MY_TABLE" V1
WHERE (CAST(V1."column_name" AS DECIMAL) = CAST(259.28 AS DECIMAL))
```
What is not clear to me is why not just compare the numbers. As I said the query without any cast returns exactly what I was expecting.

**Comment by bcogrel** [comment]
Indeed, it would be better not to have these casts. By default, DECIMAL has a scale of 0, which basically prunes the floating part (i.e. it treats it as an integer)… . Hence, your bug.

The current implementation of [DB2DBTypeFactory](
https://github.com/ontop/ontop/blob/53163dc4ee84356f30efe75a0b38af930b06295e/db/rdb/src/main/java/it/unibz/inf/ontop/model/type/impl/DB2DBTypeFactory.java) is minimalist, more work is needed to better model DB2 types. 

We may to rely on the similar trick as for Redshift, where a ["default floating decimal" type](https://github.com/ontop/ontop/blob/53163dc4ee84356f30efe75a0b38af930b06295e/db/rdb/src/main/java/it/unibz/inf/ontop/model/type/impl/RedshiftDBTypeFactory.java#L29C33-L29C52) is defined.

**Comment by marcelomachado** [comment]
Thanks @bcogrel, 
In fact it looks like a minimalist implementation, it is possible to notice in this code that you sent me the existence of some [variables created but not used](https://github.com/ontop/ontop/blob/53163dc4ee84356f30efe75a0b38af930b06295e/db/rdb/src/main/java/it/unibz/inf/ontop/model/type/impl/DB2DBTypeFactory.java#L19-L23).

I will try to come up with some solution, but could you help me with [this question](https://github.com/ontop/ontop/discussions/751)?
